Bundling kserving backend and client in the same pypi

jinchihe opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Develop python pip package for KFServing SDK, user can directly install kfserving sdk using `pip install kfserving`, and then call kfserving sdk apis to create/check/patch/delete kfservice.

**Which issue(s) this PR fixes** :
Fixes #246 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Support installing KFServing Python SDK by `pip install kfserving` 
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kfserving/250)
<!-- Reviewable:end -->

johnugeorge commented on sdk/requirements.txt:
```diff
@@ -3,3 +3,4 @@ six >= 1.10
 python_dateutil >= 2.5.3
 setuptools >= 21.0.0
 urllib3 >= 1.15.1
+kubernetes
```
version?

jinchihe replied:
Udpated, Thanks!

jinchihe commented:
The `kfserving` python SDK has been published to here: https://pypi.org/project/kfserving/

And user can install kfserving sdk from pip easily, and kfserving SDK works fine after installation.
```
(python36) [root@jinchi1 sdk]# pip install kfserving
Collecting kfserving
  Downloading https://files.pythonhosted.org/packages/71/7d/62da6cfcf79522351ac1d1327fbdcfdc7de84a75afc7596df210b69c8dab/kfserving-0.0.1.tar.gz
.....
Building wheels for collected packages: kfserving
  Building wheel for kfserving (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/f1/8d/2f/c4b8db954491f4d997178c302acc3d4e66585899d25bb2e9db
Successfully built kfserving
Installing collected packages: kfserving
Successfully installed kfserving-0.0.1
```

jinchihe commented:
/assign @yuzisun 
Thank you in advance!

jinchihe commented:
/retest

jinchihe commented:
/retest

jinchihe commented:
/retest

yuzisun commented:
mostly looks good to me, do you currently publish to PyPi manually ?

yuzisun commented on sdk/setup.py:
```diff
@@ -1,17 +1,44 @@
-NAME = "kfserving"
-VERSION = "0.1"
+setuptools.setup(
+    name='kfserving',
```
@ellis-bigelow what do you think about the name? I think we have the other package called kfserver ?

jinchihe replied:
Personally I set it to `kfserving` as want to keep consistent with out product name. :-) 
Waiting for @ellis-bigelow inputs.  Thanks.

yuzisun replied:
How about `kfserving-sdk`?

ellis-bigelow replied:
Here's a thought. We already are planning on making our KFService and KFModel libraries available on pypi. Should these be in the same library? 

It is nontrivial overhead to manage the release of multiple packages like this. I don't imagine we will have so much code that it will be prohibitive to have it in the same library. We can always split in the future as well.

What do you think?

animeshsingh replied:
kubeflow pipelines sdk is called kfp
Why not just kfs?

cliveseldon replied:
Makes sense.

jinchihe replied:
From the meeting, we planed to bundling the server and client into same pypi, and discuss more in the #268

jinchihe commented:
> mostly looks good to me, do you currently publish to PyPi manually ?

Yes, I do that manually, since there are only one command `python setup.py sdist upload -r pypi` to do that, and before running the CLI, we also need to configure the `.pypirc` for user-owned username/password, so I did not set it in script.

jinchihe commented:
/hold

We are discussing more details in the #268 , may update once somethings decided. Thanks.

jinchihe commented:
Hello Reviews:

From discussion and analysis from the ticket #268 , I update the PR changes, since the update is large, I would like to explain more for easy revewing. Overall,  I merged the `sdk` into the `python/kfserving`, the especial changes are as following:
1. Merged `python/kfserving/kfserving/__init__.py` and `sdk/kfserving/__init__.py` 
2. Merged the `python/kfserving/setup.py` and `sdk/setup.py`
3. Moved `test_server.py` and `test_storage.py` under `python/kfserving/kfserving/` to `python/kfserving/test`
4. Renamed the client `README.md` to `KFServing_client.md`, in fact I’m hesitating if that should be merged with server README, but keep separation since the README will auto generated for easy maintain.
5. Ignored autogenerated below files but I think useless: `.swagger-codegen-ignore .swagger-codegen/ .travis.yml git_push.sh test-requirements.txt tox.ini` under `sdk`

There is a problem that the `sdk-gen.sh` will overwrite some file such current `README.md` in the future, that need to be updated that manually before checkin. Maybe we can update the script to genereate to tmp dir firstlt, and update something and then copy the `python/kfserving`.

/cc @yuzisun @ellis-bigelow @animeshsingh

animeshsingh commented on python/kfserving/setup.py:
```diff
@@ -12,32 +12,58 @@
+    version='0.0.2',
+    author="Kubeflow Authors",
     author_email='ellisbigelow@google.com',
     license='../../LICENSE.txt',
```
You may want to add your own email as author/maintainer for this - so that we know who to reach out incase things don't work :-)

jinchihe replied:
Added, thanks!

animeshsingh commented on python/kfserving/setup.py:
```diff
@@ -12,32 +12,58 @@
+        'Programming Language :: Python :: 3',
+        'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3.6',
+        'Programming Language :: Python :: 3.7',
```
Don't think we support Python version before 3.6 in KFServing?

jinchihe replied:
Yes, we does not tested on Python 3.5, to be save, removed 3.5.

animeshsingh commented on python/kfserving/setup.py:
```diff
@@ -12,32 +12,58 @@
+    classifiers=[
+        'Intended Audience :: Developers',
+        'Intended Audience :: Education',
+        'Intended Audience :: Science/Research',
```
'Intended Audience :: Developers',
'Intended Audience :: Academists',
'Intended Audience :: Scientists/Researchers',

jinchihe replied:
Updated. Thanks!

jinchihe replied:
Rollbacked, since the updated ones are not valid. @animeshsingh 

```
Error: 'Intended Audience :: Academists' is not a valid
Error: 'Intended Audience :: Scientists/Researchers' is not a valid
```

jinchihe commented:
Hello @yuzisun @ellis-bigelow @animeshsingh ,

The PR is ready for reviewing now, please get chance to take a look. I summarized what's done in the PR as [above comments](https://github.com/kubeflow/kfserving/pull/250#issuecomment-515386922) for easy reviewing since the merging changes are huge :-) Thanks a lot!

I already released the KFServing Python SDK version 0.1.1 version to [Pypi](https://pypi.org/project/kfserving/), and finished some tests manully for both server and client, that works fine.

/unhold

jinchihe commented:
/hold cancel

animeshsingh commented:
CC @yuzisun

ellis-bigelow commented on hack/Python-sdk/README.md:
```diff
@@ -0,0 +1,53 @@
+
```
can we have this by /hack/python-sdk?

jinchihe replied:
@ellis-bigelow Thanks, updated to `python-sdk`.

ellis-bigelow commented on python/kfserving/README.md:
```diff
@@ -1,6 +1,28 @@
+
+
+## KFServing Server
+KFServing is a unit of model serving. KFServing's python libraries implement a standardized KFServing library that is extended by model serving frameworks such as XGBoost and PyTorch. It encapsulates data plane API definitions and storage retrieval for models.
```
"KFServing is a unit of model serving." is an awkward statement here.

jinchihe replied:
Updated, thanks.

jinchihe commented:
Rebased to resove the conficts, but presubmit faild, that's caused by no space in automation env.
```
fatal: could not create leading directories of '/mnt/test-data-volume/kubeflow-kfserving-presubmit-e2e-250-7dd4267-5664-a27c/src/kubeflow
/kfserving': No space left on device
```

animeshsingh commented:
> There is a problem that the `sdk-gen.sh` will overwrite some file such current `README.md` in the future, that need to be updated that manually before check-in.

Can we make a comment on top of sdk-gen.sh to remind that folks need to replace README.md post the sdk generation? 

Nice job @jinchihe

jinchihe commented:
> Can we make a comment on top of sdk-gen.sh to remind that folks need to replace README.md post the sdk generation?

@animeshsingh Great comments!  Added notes in the scripts and README file for the script.

jinchihe commented:
/retest

ellis-bigelow commented:
Nice!
/lgtm
/approve

jinchihe commented:
/retest

jinchihe commented:
/retest

animeshsingh commented:
/lgtm

animeshsingh commented:
@yuzisun @ellis-bigelow final comments on this?

jinchihe commented:
> @yuzisun @ellis-bigelow final comments on this?

@animeshsingh The PR has been approved by Ellis, and That's OK from Dan, but the presummit testing already failed since timeout or some strange reasons...

/retest
